src/theory/bags/solver_state.h

```diff
@@ -31,24 +31,47 @@ class SolverState : public TheoryState
  public:
   SolverState(context::Context* c, context::UserContext* u, Valuation val);
 
-  void registerClass(TNode n);
+  void registerBag(TNode n);
 
-  Node registerBagElement(TNode n);
+  /**
+   * @param n has the form (bag.count e A)
+   * @pre bag A needs is already registered using registerBag(A)
+   * @return a unique skolem for (bag.count e A)
+   */
+  Node registerCountTerm(TNode n);
+  /** get all bag terms */
+  const std::set<Node>& getBags();
+  /**
+   * @pre B is a registered bag
+   * @return all elements associated with bag B so far
+   * Note that associated elements are not necessarily elements in B
+   * Example:
+   * (assert (= 1 (bag.count x (difference_remove A B))))
+   * element x is associated with bags A, B, (difference_remove A B)
+   * albeit x is definitely not in B.
+   */
+  const std::set<Node>& getElements(Node B);
+  /**
+   * @param countTerm has the form (bag.count e A)
+   * @return a unique skolem for countTerm
+   */
+  Node getCountSkolem(const Node& countTerm);
 
-  std::set<Node>& getBags();
+  /** clear all bags data structures */
+  void reset();
 
-  std::set<Node>& getElements(TypeNode t);
-
-  std::map<Node, Node>& getBagElements(Node B);
+  /** merge the elements of the two registered bags n1, n2 */
```

Comment:
I'm confused why you are implementing this method.

I thought the elements of a bag were only being computed at full effort? 

This method is called prior to full effort, when equivalence classes merge.  You should not be implementing something like this (or if you were, the data structures in this class would all have to be SAT-context-dependent and the design of this class would have to change considerably).

For now, I do not think this method will be necessary.

Reply:
Initially I started collected all terms at post check, and then ran into the issue of merging elements in an equivalent class. 
```
[Z , (union_max A B), (intersection_min A B)] and 
terms (count x Z), (count y (intersection_min A B))
```
I thought it is simpler to use notifications for now and worry about context-dependence later.